src/iceberg/type.h

```diff
@@ -0,0 +1,411 @@
+
+  TypeId type_id() const override;
+  std::string ToString() const override;
+
+ protected:
+  bool Equals(const Type& other) const override;
+};
+
+/// \brief A data type representing a wall clock time in microseconds without
+///   reference to a timezone or date.
+class ICEBERG_EXPORT TimeType : public PrimitiveType {
+ public:
+  TimeType() = default;
+  ~TimeType() = default;
+
+  TypeId type_id() const override;
+  std::string ToString() const override;
+
+ protected:
+  bool Equals(const Type& other) const override;
+};
+
+/// \brief A base class for any timestamp time (irrespective of unit or
+///   timezone).
+class ICEBERG_EXPORT TimestampBase : public PrimitiveType {
+ public:
+  /// \brief Is this type zoned or naive?
+  [[nodiscard]] virtual bool is_zoned() const = 0;
+  /// \brief The time resolution.
+  [[nodiscard]] virtual TimeUnit time_unit() const = 0;
+};
+
+/// \brief A data type representing a timestamp in microseconds without
+///   reference to a timezone.
+class ICEBERG_EXPORT TimestampType : public TimestampBase {
+ public:
+  TimestampType() = default;
+  ~TimestampType() = default;
+
+  bool is_zoned() const override;
+  TimeUnit time_unit() const override;
+
+  TypeId type_id() const override;
+  std::string ToString() const override;
+
+ protected:
+  bool Equals(const Type& other) const override;
+};
+
+/// \brief A data type representing a timestamp as microseconds since the
+///   epoch in UTC.
+class ICEBERG_EXPORT TimestampTzType : public TimestampBase {
+ public:
+  TimestampTzType() = default;
+  ~TimestampTzType() = default;
+
+  bool is_zoned() const override;
+  TimeUnit time_unit() const override;
+
+  TypeId type_id() const override;
+  std::string ToString() const override;
+
+ protected:
+  bool Equals(const Type& other) const override;
+};
+
+/// \brief A data type representing a bytestring.
+class ICEBERG_EXPORT BinaryType : public PrimitiveType {
+ public:
+  BinaryType() = default;
+  ~BinaryType() = default;
+
+  TypeId type_id() const override;
+  std::string ToString() const override;
+
+ protected:
+  bool Equals(const Type& other) const override;
+};
+
+/// \brief A data type representing a string.
+class ICEBERG_EXPORT StringType : public PrimitiveType {
+ public:
+  StringType() = default;
+  ~StringType() = default;
+
+  TypeId type_id() const override;
+  std::string ToString() const override;
+
+ protected:
+  bool Equals(const Type& other) const override;
+};
+
+/// \brief A data type representing a fixed-length bytestring.
+class ICEBERG_EXPORT FixedType : public PrimitiveType {
+ public:
+  /// \brief Construct a fixed type with the given length.
+  FixedType(int32_t length);
+  ~FixedType() = default;
+
+  /// \brief The length (the number of bytes to store).
+  [[nodiscard]] int32_t length() const;
+
+  TypeId type_id() const override;
+  std::string ToString() const override;
+
+ protected:
+  bool Equals(const Type& other) const override;
+
+ private:
+  int32_t length_;
```

Comment:
iceberg-rust has the length of `u64`[1], `int32_t` can represent about 2 gigabytes, maybe we should change to `int64_t` to be compatible with other language implementation?

[1] https://github.com/apache/iceberg-rust/blob/main/crates/iceberg/src/spec/datatypes.rs#L241

Reply:
Java uses a 32 bit int. https://github.com/apache/iceberg/blob/d693f83b6ad6e306550d117ff0bf0014356500b2/api/src/main/java/org/apache/iceberg/types/Types.java#L360

Are the Iceberg maintainers interested in making the spec actually lay out these details? I've mostly copied Java since the spec itself is extremely vague.

Reply:
The int32 is fine. No one should go beyond 2GB values for a single value.